src/event.js

```diff
@@ -574,55 +531,63 @@ jQuery.event = {
 // already occurred before other listeners are invoked.
-function leverageNative( el, type, forceAdd, allowAsync ) {
+function leverageNative( el, type, expectSync ) {
 
-	// Setup must go through jQuery.event.add
-	if ( forceAdd ) {
-		jQuery.event.add( el, type, forceAdd );
+	// Missing expectSync indicates a trigger call, which must force setup through jQuery.event.add
+	if ( !expectSync ) {
+		jQuery.event.add( el, type, returnTrue );
 		return;
 	}
 
 	// Register the controller as a special universal handler for all event namespaces
-	dataPriv.set( el, type, forceAdd );
+	dataPriv.set( el, type, false );
 	jQuery.event.add( el, type, {
 		namespace: false,
 		handler: function( event ) {
-			var maybeAsync, result,
+			var notAsync, result,
 				saved = dataPriv.get( this, type );
 
-			// Interrupt processing of the outer synthetic .trigger()ed event
-			if ( ( event.isTrigger & 1 ) && this[ type ] && !saved ) {
-
-				// Store arguments for use when handling the inner native event
-				saved = slice.call( arguments );
-				dataPriv.set( this, type, saved );
-
-				// Trigger the native event and capture its result
-				// Support: IE <=9 - 11+
-				// focus() and blur() are asynchronous
-				maybeAsync = allowAsync( this, type );
-				this[ type ]();
-				result = dataPriv.get( this, type );
-				if ( result !== saved ) {
-					dataPriv.set( this, type, false );
-
-					// Cancel the outer synthetic event
-					event.stopImmediatePropagation();
-					event.preventDefault();
-					return result;
-				} else if ( maybeAsync ) {
-
-					// Cancel the outer synthetic event in expectation of a followup
-					event.stopImmediatePropagation();
-					event.preventDefault();
-					return;
-				} else {
-					dataPriv.set( this, type, false );
+			if ( ( event.isTrigger & 1 ) && this[ type ] ) {
+
+				// Interrupt processing of the outer synthetic .trigger()ed event
+				if ( !saved ) {
+
+					// Store arguments for use when handling the inner native event
+					saved = slice.call( arguments );
+					dataPriv.set( this, type, saved );
+
+					// Trigger the native event and capture its result
+					// Support: IE <=9 - 11+
+					// focus() and blur() are asynchronous
+					notAsync = expectSync( this, type );
+					this[ type ]();
+					result = dataPriv.get( this, type );
+					if ( saved !== result || notAsync ) {
+						dataPriv.set( this, type, false );
+					} else {
+						result = undefined;
+					}
+					if ( saved !== result ) {
+
+						// Cancel the outer synthetic event
+						event.stopImmediatePropagation();
+						event.preventDefault();
+						return result;
+					}
+
+				// If this is an inner synthetic event for an event with a bubbling surrogate
+				// (focus or blur), assume that the surrogate already propagated from triggering the
+				// native event and prevent that from happening again here.
+				// This technically gets the ordering wrong w.r.t. to `.trigger()`, but that seems
```

Comment:
Could you explain where the order is wrong considering the `focusin`/`focusout` ordering test passes?

Reply:
Updated the comment. It is subtle (and I may actually be wrong about it), but it will change with #4300 anyway.